npeps/npep-173-service-account-matches.md

```diff
@@ -0,0 +1,99 @@
+for policies.
+
+## User-Stories/Use-Cases
+
+1. **As a** cluster admin **I want** to select pods using their service accounts
+   instead of labels **so that** I can avoid having to setup webhooks and
+   validation admission policies to prevent users from changing labels on their
+   namespaces and pods that will make my policy start/stop matching the traffic
+   which is undesired
+2. **As a** cluster admin **I want** to select pods using their service accounts
+   instead of labels **so that** I can avoid the scale impact caused due to
+   mutation of pod/namespace labels that will cause a churn which makes my CNI
+   implementation react to that every time user changes a label.
+3. **As a** cluster admin my workloads have immutable identities and **I want**
+   to apply policies per workloads using their service accounts instead of labels
+   **since** I want to have eventual consistency of that policy in the cluster.
+
+## Unresolved Discussions
+
+* How to provide a standard way to configure/describe the service mesh behavior
+  of intercepting traffic and deciding whether to allow it based on information
+  in the TLS packets?
+  * NetworkPolicies apply on L3 and L4 while Meshes operate at L7 mostly. So when
+    trying to express "denyAll connections except the allowed serviceAccount to
+    serviceAccount connections"; how do we split the enforcement in this scenario
+    between the CNI plugin that implementations the network policy at L3/L4 and
+    service mesh implementation that implements the policy at L7?
+    * One way is to probably split the implementation responsibility:
+      1. CNI plugin can take of implementing denyAll connections and
+         allowed serviceAccount to serviceAccount connections upto L3/L4
+      2. service mesh implementation can implement the allowed serviceAccount to
+         serviceAccount connections on L7
+    * NOTE: There are some service mesh implementations which have a
+      CNI component in them that collapses the logic of enforcement - so maybe
+      that might become a requirement that the implementation be able to handle
+      end to end enforcement of the full policy
+  * We might also want to express additional constraints at L7 (such as only
+    allowing requests from a given identity on specific paths or with specific
+    HTTP/gRPC methods) - Ideally we would want these extra fields to be coexisting
+    with the L3/L4 Authorization policy
+* Should this really be part of AdminNetworkPolicy and NetworkPolicy APIs or should
+  this be a new CRD?
+  * Making it part of existing APIs: Existing network policy APIs are pretty heavy
+    with other types of selectors like namespaces, pods, nodes, networks, fqdns and
+    expecting mesh implementations to implement the full CRD just to get the identity
+    based selection might not be practical
+  * Making it part of new API: We fall back to compatibility problem of the different
```

Comment:
why compatibility issues, the different policies applies should be an AND, in which sense is this incompatible?

Reply:
I dont see compatibility issues but I can see terrible UX if we end up with another API which sits *on top* of netpol. This means that as a user, to allow point A to talk with point B on a cluster with default deny I need to:
  1. add a netpol to allow the traffic
  2. add an identity-based policy to allow it
  3. potentially add an L7 policy if I want L7 attributes based policy.

Reply:
I think there are a non-trivial set of implementations that would like to support an identity-based version of network policy, but would not support the additional set of APIs based on label selectors. To me, that suggests that it would be cleaner to have a new API focused on identity-based APIs. 

I agree that @LiorLieberman's example is not a great UX, but I'd hope that most users would be fully served by only one of these APIs and would not need to layer like that. Could we express default deny within this new policy to avoid layering?

Reply:
FWIW in Istio we would ideally have a policy that consists only of SA/identity and not pod label selectors.

Reply:
> why compatibility issues, the different policies applies should be an AND, in which sense is this incompatible?

probably compatibility is not the right word here, but I meant the layering problem, but if we agree "AND" of the APIs is what we are after wherein L3/L4 netpol API is denying a connection but the new identity API one wants to allow that (for podsSAs) then we deny cause we need it to be allow on all APIs and there is one API denying the connection even if the other API explicitly allows it?

Reply:
> Could we express default deny within this new policy to avoid layering?
@aojea interested in your thoughts regarding default deny with this